core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -0,0 +1,719 @@
+            maybeEpoch = leaderEpochCache.nextEpoch(epoch);
+        }
+
+        return Optional.empty();
+    }
+
+    private static abstract class CancellableRunnable implements Runnable {
+        private volatile boolean cancelled = false;
+
+        public void cancel() {
+            cancelled = true;
+        }
+
+        public boolean isCancelled() {
+            return cancelled;
+        }
+    }
+
+    /**
+     * Returns the leader epoch checkpoint by truncating with the given start[exclusive] and end[inclusive] offset
+     *
+     * @param log         The actual log from where to take the leader-epoch checkpoint
+     * @param startOffset The start offset of the checkpoint file (exclusive in the truncation).
+     *                    If start offset is 6, then it will retain an entry at offset 6.
+     * @param endOffset   The end offset of the checkpoint file (inclusive in the truncation)
+     *                    If end offset is 100, then it will remove the entries greater than or equal to 100.
+     * @return the truncated leader epoch checkpoint
+     */
+    InMemoryLeaderEpochCheckpoint getLeaderEpochCheckpoint(UnifiedLog log, long startOffset, long endOffset) {
+        InMemoryLeaderEpochCheckpoint checkpoint = new InMemoryLeaderEpochCheckpoint();
+        if (log.leaderEpochCache().isDefined()) {
+            LeaderEpochFileCache cache = log.leaderEpochCache().get().writeTo(checkpoint);
+            if (startOffset >= 0) {
+                cache.truncateFromStart(startOffset);
+            }
+            cache.truncateFromEnd(endOffset);
+        }
+
+        return checkpoint;
+    }
+
+    class RLMTask extends CancellableRunnable {
+
+        private final TopicIdPartition topicIdPartition;
+        private final Logger logger;
+
+        private volatile int leaderEpoch = -1;
+
+        public RLMTask(TopicIdPartition topicIdPartition) {
+            this.topicIdPartition = topicIdPartition;
+            LogContext logContext = new LogContext("[RemoteLogManager=" + brokerId + " partition=" + topicIdPartition + "] ");
+            logger = logContext.logger(RLMTask.class);
+        }
+
+        boolean isLeader() {
+            return leaderEpoch >= 0;
+        }
+
+        // The copiedOffsetOption is OptionalLong.empty() initially for a new leader RLMTask, and needs to be fetched inside the task's run() method.
+        private volatile OptionalLong copiedOffsetOption = OptionalLong.empty();
+
+        public void convertToLeader(int leaderEpochVal) {
+            if (leaderEpochVal < 0) {
+                throw new KafkaException("leaderEpoch value for topic partition " + topicIdPartition + " can not be negative");
+            }
+            if (this.leaderEpoch != leaderEpochVal) {
+                leaderEpoch = leaderEpochVal;
+            }
+            // Reset readOffset, so that it is set in next run of RLMTask
+            copiedOffsetOption = OptionalLong.empty();
```

Comment:
Should the offset be reset if the leader epoch is the same as the one we already have? (it shouldn't hurt to query the highest remote offset multiple times in any case, so that is a minor observation).